agora/Agora/Proposal/Scripts.hs

```diff
@@ -368,5 +376,89 @@ proposalValidator proposal =
+
+          let -- Only the status of proposals should be updated in this case.
+              templateProposalOut =
+                mkRecordConstr
+                  PProposalDatum
+                  ( #proposalId .= proposalF.proposalId
+                      .& #effects .= proposalF.effects
+                      .& #status .= proposalOutStatus
+                      .& #cosigners .= proposalF.cosigners
+                      .& #thresholds .= proposalF.thresholds
+                      .& #votes .= proposalF.votes
+                      .& #timingConfig .= proposalF.timingConfig
+                      .& #startingTime .= proposalF.startingTime
+                  )
+
+          tcassert "Only status changes in the output proposal" $
+            templateProposalOut #== proposalOut
+
+          inDraftPeriod <- tclet $ isDraftPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime
+          inVotingPeriod <- tclet $ isVotingPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime
+          inLockedPeriod <- tclet $ isLockingPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime
+          inExecutionPeriod <- tclet $ isExecutionPeriod # proposalF.timingConfig # proposalF.startingTime # currentTime
+
+          -- Check the timings.
+          let isFinished = proposalF.status #== pconstantData Finished
+
+              notTooLate = pmatch (pfromData proposalF.status) $ \case
+                PDraft _ -> inDraftPeriod
+                PVotingReady _ -> inVotingPeriod
+                PLocked _ -> inExecutionPeriod
+                _ -> pconstant False
+
+          tcassert "Finished proposals cannnot be advanced" $ pnot # isFinished
+
+          pure $
+            pif
+              notTooLate
+              -- On time: advance to next status.
+              ( pmatch (pfromData proposalF.status) $ \case
+                  PDraft _ -> unTermCont $ do
+                    -- TODO: Perform other necessary checks.
+
+                    -- 'Draft' -> 'VotingReady'
+                    tcassert "Proposal status set to VotingReady" $
+                      proposalOutStatus #== pconstantData VotingReady
+
+                    pure $ popaque (pconstant ())
+                  PVotingReady _ -> unTermCont $ do
+                    -- 'VotingReady' -> 'Locked'
+                    tcassert "Proposal status set to Locked" $
+                      proposalOutStatus #== pconstantData Locked
+
+                    -- Check that the highest votes meet the minimum requirement.
+                    let winner = mustBePJust # "Highest votes not found" #$ pwinner # proposalF.votes
+                        highestVotes = pfromData $ psndBuiltin # winner
+
+                    tcassert "Highest vote count should exceed the threshold" $
+                      pto (pto $ pfromData thresholdsF.execute) #< highestVotes
+
+                    pure $ popaque (pconstant ())
```

Comment:
Adding to my other comment: If the timing is right -- the voting period is over -- then it should _always_ be possible to go to the `Locked` status. This is as a result of the winner being deterministic and unambiguous. Additionally, as a result of this, the assertion of `highestVotes` here, should be performed in `pwinner`, I suppose?

Though, `pwinner` is implemented in a nice way currently. Maybe we need to rename it to `phighestVotes` and then have a `pwinner` which adds these bits of logic, if that makes sense.

There are some inconsistencies in the documentation/naming that I can fix after this PR gets merged.

Reply:
Yeah, this is way more reasonable. Also, I think we need more doc reguarding the timing.